index.ts

```diff
@@ -0,0 +1,190 @@
+ */
+
+import optionsConv = require('./lib/option_converter')
+import Agent from './lib/agent'
+import Server from './lib/server'
+import IncomingMessage from './lib/incoming_message'
+import OutgoingMessage from './lib/outgoing_message'
+import ObserveReadStream = require('./lib/observe_read_stream')
+import ObserveWriteStream = require('./lib/observe_write_stream')
+import { parameters, refreshTiming, defaultTiming } from './lib/parameters'
+import { CoapMethod, OptionName, Packet } from 'coap-packet'
+import { Socket } from 'dgram'
+import { isIPv6 } from 'net'
+
+export type OptionValue = string | number | Buffer | Buffer[]
+export type BlockCacheMap<T> = Map<string, {payload: T, timeoutId: NodeJS.Timeout}>
+
+export interface Option {
+    name: number | OptionName | string
+    value: OptionValue
+}
+
+export interface Block {
+    more: number
+    num: number
+    size: number
+}
+
+export function createServer (options?: CoapServerOptions | typeof requestListener, listener?: typeof requestListener): Server {
+    return new Server(options, listener)
+}
+
+function _getHostname (url: URL): string {
+    const hostname = url.hostname
+    // Remove brackets from literal IPv6 addresses
+    if (hostname.startsWith('[') && hostname.endsWith(']')) {
+        return hostname.substring(1, hostname.length - 1)
+    }
+    return hostname
+}
+
+function _getQueryParamsFromSearch (url: URL): string | undefined {
+    if (url.search != null) {
+        return url.search.substring(1)
+    }
+}
+
+function _getPort (url: URL): number {
+    if (url.port !== '') {
+        return parseInt(url.port)
+    } else {
+        return parameters.coapPort
+    }
+}
+
+function _parseUrl (url: string): CoapRequestParams {
+    const requestParams: CoapRequestParams = {}
+    const parsedUrl = new URL(url)
+    requestParams.hostname = _getHostname(parsedUrl)
+    requestParams.query = _getQueryParamsFromSearch(parsedUrl)
+    requestParams.port = _getPort(parsedUrl)
+    requestParams.pathname = parsedUrl.pathname
+    return requestParams
+}
+
+export function request (requestParams: CoapRequestParams | string): OutgoingMessage {
+    let agent: Agent
+    if (typeof requestParams === 'string') {
+        requestParams = _parseUrl(requestParams)
+    }
+    const ipv6 = isIPv6(requestParams.hostname ?? requestParams.host ?? '')
+    if (requestParams.agent != null && requestParams.agent !== false) {
+        agent = requestParams.agent
+    } else if (requestParams.agent === false && !ipv6) {
+        agent = new Agent({ type: 'udp4' })
+    } else if (requestParams.agent === false && ipv6) {
+        agent = new Agent({ type: 'udp6' })
+    } else if (ipv6) {
+        agent = exports.globalAgentIPv6
+    } else {
+        agent = exports.globalAgent
+    }
+    return agent.request(requestParams)
+}
+
+export type CoapOptions = Partial<Record<OptionName, OptionValue>>
+
+export interface CoapPacket extends Packet {
```

Comment:
Can you move these interfaces to their own files in a domain/models folder and import them it would clean up this file quite a bit?

Reply:
Thank you, that is a very good point! I will soon push a commit addressing your suggestion :)

Reply:
Good catch! Should be fixed now :)